<center><b>Normal Distribution and Z Score: Math and statistics for data science</b>


In [13]:
import pandas as pd
import seaborn as sn

We are going to use heights dataset from kaggle.com. Dataset has heights and weights both but I have removed weights to make it simple

https://www.kaggle.com/mustafaali96/weight-height

In [14]:
df = pd.read_csv("heights.csv")
df.head()

gender     height
0   Male  73.847017
1   Male  68.781904
2   Male  74.110105
3   Male  71.730978
4   Male  69.881796

<b>Outlier detection and removal using Standard Deviation

In [16]:
df.height.describe()

count    10000.000000
mean        66.367560
std          3.847528
min         54.263133
25%         63.505620
50%         66.318070
75%         69.174262
max         78.998742
Name: height, dtype: float64

In [24]:
std_deviation = df.height.std()
std_deviation

3.847528120795573

In [25]:
mean-3*std_deviation

54.824975392478876

In [26]:
mean+3*std_deviation

77.91014411725232

In [27]:
df[(df.height < 54.82) | (df.height > 77.91)]

gender     height
994     Male  78.095867
1317    Male  78.462053
2014    Male  78.998742
3285    Male  78.528210
3757    Male  78.621374
6624  Female  54.616858
9285  Female  54.263133

In [28]:
df_no_outlier = df[(df.height<77.91) & (df.height>54.82)]
df_no_outlier.shape

(9993, 2)

<b>Outlier detection and removal using Z Score</b>

Z score is a way to achieve same thing that we did above 

Z score indicates how many standard deviation away a data point is.

For example in our case mean is 66.37 and standard deviation is 3.84.

If a value of a data point is 77.91 then Z score for that is 3 because it is 3 standard deviation away (77.91 = 66.37 + 3 * 3.84)

Calculate the Z Score
<img src = "zscore">


Let's add a new column in our dataframe for this Z score

In [29]:
df['zscore'] = ( df.height - df.height.mean() ) / df.height.std()
df.head(5)

gender     height    zscore
0   Male  73.847017  1.943964
1   Male  68.781904  0.627505
2   Male  74.110105  2.012343
3   Male  71.730978  1.393991
4   Male  69.881796  0.913375

Above for first record with height 73.84, z score is 1.94. This means 73.84 is 1.94 standard deviation away from mean

In [30]:
df.height.mean()

66.3675597548656

In [31]:
df.height.std()

3.847528120795573

In [32]:
(73.84-66.37)/3.84

1.9453124999999998

In [33]:
df[df['zscore']>3]

gender     height    zscore
994    Male  78.095867  3.048271
1317   Male  78.462053  3.143445
2014   Male  78.998742  3.282934
3285   Male  78.528210  3.160640
3757   Male  78.621374  3.184854

In [34]:
df[df['zscore']<-3]


gender     height    zscore
6624  Female  54.616858 -3.054091
9285  Female  54.263133 -3.146027